### PR TITLE
Adding simple battery subsystem and Dymos integration test

### PR DESCRIPTION
**Pull request by johnjasa**
### Summary

This builds upon @jkirk5's initial work to get a simple battery model into Aviary.
I added in a simple Dymos-based mission with the battery model and an accompanying test.
This is primarily to enable electroprop aircraft but has general uses.

### Related Issues

- Resolves #

### Backwards incompatibilities

None

### New Dependencies

None

### CODE REVIEW

File: aviary/models/engines/turbofan_28k_with_electric.deck
Comment (crecine) [inline aviary/models/engines/turbofan_28k_with_electric.deck:1]: is this electric power in or out?
e-boost vs generator

File: aviary/subsystems/energy/battery_builder.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+from aviary.subsystems.subsystem_builder_base import SubsystemBuilderBase
+from aviary.subsystems.energy.battery_sizing import SizeBattery
+from aviary.variable_info.variables import Aircraft, Dynamic
+
+
+class BatteryBuilder(SubsystemBuilderBase):
+    def build_pre_mission(self, aviary_inputs=None):
+        return SizeBattery(aviary_inputs=aviary_inputs)
+
+    def get_mass_names(self):
```
Comment (crecine) [inline aviary/subsystems/energy/battery_builder.py:13]: should this be a property?

File: aviary/subsystems/energy/battery_builder.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        return [Aircraft.Battery.MASS]
+
+    def build_mission(self, num_nodes, aviary_inputs=None) -> om.Group:
+        battery_group = om.Group()
+        soc = om.ExecComp('state_of_charge = (energy_capacity - (electric_energy/efficiency)) / energy_capacity',
+                          state_of_charge={'val': np.zeros(
```
Comment (crecine) [inline aviary/subsystems/energy/battery_builder.py:19]: is the storage efficiency, charge efficiency, discharge efficiency, etc?

File: aviary/subsystems/energy/battery_builder.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    def get_constraints(self):
+        # discharge_limit = aviary_options.get_val(Aircraft.Battery.DISCHARGE_LIMIT)
+        # constraint_dict = {f'path_{Dynamic.Mission.BATTERY_STATE_OF_CHARGE} = {Dynamic.Mission.BATTERY_STATE_OF_CHARGE} - {Aircraft.Battery.DISCHARGE_LIMIT}':
+        #                    {'type': 'path',
+        #                     'lower': 0.0}}
+        constraint_dict = {
+            # Dynamic.Mission.BATTERY_STATE_OF_CHARGE:
+            #                {'type': 'path',
+            #                 'lower': 0.2},
+        }
```
Comment (crecine) [inline aviary/subsystems/energy/battery_builder.py:56]: remove?

File: aviary/subsystems/energy/test/test_battery.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        prob.setup(force_alloc_complex=True)
+
+        prob.set_val(av.Aircraft.Battery.PACK_ENERGY_DENSITY, 550, units='kW*h/kg')
+        prob.set_val(av.Aircraft.Battery.PACK_MASS, 1200, units='lbm')
+        prob.set_val(av.Aircraft.Battery.ADDITIONAL_MASS, 115, units='lbm')
+
+        prob.run_model()
+
+        mass_expected = 1_315
+        energy_expected = 1_077_735_471.12
```
Comment (crecine) [inline aviary/subsystems/energy/test/test_battery.py:35]: maybe do this in MJ so that the number isn't quite so big

File: aviary/subsystems/test/subsystem_tester.py [hunk 5]
```diff
@@ -262,14 +261,6 @@ def test_check_state_variables(self):
                 self.assertTrue(state_var_exists,
                                 f"State variable '{key}' not found in the model.")
 
-                # Check rate_source variable existence
-                if 'rate_source' in value:
-                    rate_source = value['rate_source']
-                    rate_source_exists = any(
-                        rate_source in output[0] for output in outputs)
-                    self.assertTrue(
-                        rate_source_exists, f"Rate source variable '{rate_source}' for state '{key}' not found in the model.")
```
Comment (crecine) [inline aviary/subsystems/test/subsystem_tester.py:271]: why was this removed?

File: aviary/variable_info/variable_meta_data.py [hunk 6]
```diff
@@ -203,6 +203,122 @@
...
+add_meta_data(
+    Aircraft.Battery.DISCHARGE_LIMIT,
+    meta_data=_MetaData,
+    historical_name={"GASP": None,
+                     "FLOPS": None,
+                     "LEAPS1": 'aircraft.inputs.L0_battery.depth_of_discharge'
+                     },
+    units='unitless',
+    desc='default constraint on how far the battery can discharge, as a percentage of '
+         'total energy capacity'
```
Comment (crecine) [inline aviary/variable_info/variable_meta_data.py:236]: a default of 0 or 20% would be reasonable

File: aviary/variable_info/variable_meta_data.py [hunk 7]
```diff
@@ -203,6 +203,122 @@
...
+)
+
+add_meta_data(
+    Aircraft.Battery.PACK_VOLUMETRIC_DENSITY,
+    meta_data=_MetaData,
+    historical_name={"GASP": None,
+                     "FLOPS": None,
+                     "LEAPS1": None
+                     },
+    units='kW*h/kg',
```
Comment (crecine) [inline aviary/variable_info/variable_meta_data.py:305]: Wh/L is the most common unit I've seen

File: aviary/variable_info/variables.py [hunk 8]
```diff
@@ -583,11 +594,13 @@ class Mission:
         ALTITUDE = 'altitude'
         ALTITUDE_RATE = 'altitude_rate'
         ALTITUDE_RATE_MAX = 'altitude_rate_max'
+        BATTERY_STATE_OF_CHARGE = 'battery_state_of_charge'
         DENSITY = 'density'
         DISTANCE = 'distance'
         DISTANCE_RATE = 'distance_rate'
         DRAG = 'drag'
         DYNAMIC_PRESSURE = 'dynamic_pressure'
+        ELECTRIC_ENERGY = 'electric_energy'
```
Comment (crecine) [inline aviary/variable_info/variables.py:603]: Since energy isn't really an instantaneous value, this should probably have a more descriptive name like `total_electrical_energy_used` or something

**Review by crecine — COMMENTED** [review]
there might be some equivalent GASP variables that should be in the metadata, let me find those for you

File: aviary/models/engines/turbofan_28k_with_electric.deck
Comment (jkirk5) [reply aviary/models/engines/turbofan_28k_with_electric.deck:1]: The intent on the engine deck design side was that fuel flow & electric power have the same interpretation, that is positive values are consumption (to match how engine decks usually work)

Comment (jkirk5) [inline aviary/models/engines/turbofan_28k_with_electric.deck:1]: What is the source for this data?

File: aviary/variable_info/variable_meta_data.py (hunk 7 above)
Comment (jkirk5) [reply aviary/variable_info/variable_meta_data.py:305]: Yes, this currently has units for mass-specific energy. I don't have a preference for kW\*h vs. W*h for the numerator

File: aviary/variable_info/variables.py (hunk 8 above)
Comment (jkirk5) [reply aviary/variable_info/variables.py:603]: `total_mission_electric_energy`? `electric_energy_consumption`
Just throwing out some more suggestions

File: aviary/validation_cases/benchmark_tests/test_battery_in_a_mission.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+        prob.setup()
+
+        prob.set_initial_guesses()
+
+        prob.set_val(av.Aircraft.Battery.PACK_ENERGY_DENSITY, 550, units='kJ/kg')
+        prob.set_val(av.Aircraft.Battery.PACK_MASS, 1000, units='lbm')
+        prob.set_val(av.Aircraft.Battery.ADDITIONAL_MASS, 115, units='lbm')
+
+        prob.run_aviary_problem()
```
Comment (jkirk5) [inline aviary/validation_cases/benchmark_tests/test_battery_in_a_mission.py:83]: The test runs without errors, but do we know if the results are reasonable or garbage?

File: aviary/subsystems/energy/battery_builder.py (hunk 2 above)
Comment (jkirk5) [reply aviary/subsystems/energy/battery_builder.py:19]: When originally making this model I didn't go into that level of detail and just added in a universal efficiency factor so at least there is something for users to play with. I'd argue if you need that level of detail you should start looking into higher-fidelity battery models such as the external subsystem one

File: aviary/subsystems/energy/battery_builder.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    def build_mission(self, num_nodes, aviary_inputs=None) -> om.Group:
+        battery_group = om.Group()
+        soc = om.ExecComp('state_of_charge = (energy_capacity - (electric_energy/efficiency)) / energy_capacity',
+                          state_of_charge={'val': np.zeros(
+                              num_nodes), 'units': 'unitless'},
+                          energy_capacity={'val': 10.0, 'units': 'kJ'},
+                          electric_energy={'val': np.zeros(num_nodes), 'units': 'kJ'},
+                          efficiency={'val': 0.95, 'units': 'unitless'})
+
+        battery_group.add_subsystem('state_of_charge',
```
Comment (jkirk5) [inline aviary/subsystems/energy/battery_builder.py:25]: Currently it looks like SoC is getting calculated but never constrained - is the expectation that users must always manually specify they want that constraint active, rather than having the subsystem builder set it up?

**Review by jkirk5 — CHANGES_REQUESTED** [review]

File: aviary/subsystems/energy/battery_builder.py (hunk 2 above)
Comment (crecine) [reply aviary/subsystems/energy/battery_builder.py:19]: maybe just add a comment here then, something like "here efficiency is used an an average (or overall) efficiency"

File: aviary/models/engines/turbofan_28k_with_electric.deck
Comment (crecine) [reply aviary/models/engines/turbofan_28k_with_electric.deck:1]: could we change this to `electric_power_in` then to disambiguate?
NOx is positive out and generators are (currently) more common than e-boost

File: aviary/variable_info/variable_meta_data.py [hunk 11]
```diff
@@ -203,6 +203,122 @@
...
+                     },
+    units='unitless',
+    desc='default constraint on how far the battery can discharge, as a percentage of '
+         'total energy capacity'
+)
+
+add_meta_data(
+    Aircraft.Battery.EFFICIENCY,
+    meta_data=_MetaData,
+    historical_name={"GASP": None,
```
Comment (crecine) [inline aviary/variable_info/variable_meta_data.py:242]:
```suggestion
    historical_name={"GASP": 'INGASP.EFF_BAT',
```

File: aviary/variable_info/variable_meta_data.py [hunk 12]
```diff
@@ -203,6 +203,122 @@
...
+                     },
+    units='lbm',
+    desc='total mass of the battery',
+    default_value=0.0,
+)
+
+add_meta_data(
+    Aircraft.Battery.PACK_ENERGY_DENSITY,
+    meta_data=_MetaData,
+    historical_name={"GASP": None,
```
Comment (crecine) [inline aviary/variable_info/variable_meta_data.py:276]:
```suggestion
    historical_name={"GASP": 'INGASP.ENGYDEN',
```

File: aviary/variable_info/variable_meta_data.py [hunk 13]
```diff
@@ -203,6 +203,122 @@
...
+                     "LEAPS1": None
+                     },
+    units='kJ',
+    desc="total energy the battery can store"
+)
+
+add_meta_data(
+    Aircraft.Battery.MASS,
+    meta_data=_MetaData,
+    historical_name={"GASP": None,
```
Comment (crecine) [inline aviary/variable_info/variable_meta_data.py:264]:
```suggestion
    historical_name={"GASP": ['INGASP.WBATTIN', 'WBATTAVL'],
```

Comment (crecine) [reply aviary/variable_info/variable_meta_data.py:264]: we could change this to just `INGASP.WBATTIN`

File: aviary/variable_info/variable_meta_data.py [hunk 14]
```diff
@@ -6085,6 +6212,17 @@
...
 
+add_meta_data(
+    Dynamic.Mission.ELECTRIC_ENERGY,
+    meta_data=_MetaData,
+    historical_name={"GASP": None,
+                     "FLOPS": None,
+                     "LEAPS1": None
+                     },
+    units='kJ',
+    desc='Total amount of electric energy consumed by the vehicle across the mission',
```
Comment (crecine) [inline aviary/variable_info/variable_meta_data.py:6223]: is this across the whole mission, or just up until the current time?

File: aviary/variable_info/variable_meta_data.py [hunk 15]
```diff
@@ -203,6 +203,122 @@
...
+                     "LEAPS1": None
+                     },
+    units='unitless',
+    desc="battery pack efficiency"
+)
+
+add_meta_data(
+    Aircraft.Battery.ENERGY_CAPACITY,
+    meta_data=_MetaData,
+    historical_name={"GASP": None,
```
Comment (crecine) [inline aviary/variable_info/variable_meta_data.py:253]:
```suggestion
    historical_name={"GASP": 'EBATTAVL',
```

File: aviary/variable_info/variable_meta_data.py [hunk 16]
```diff
@@ -203,6 +203,122 @@
...
+                     },
+    units='lbm',
+    desc='mass of non energy-storing parts of the battery',
+    default_value=0.0,
+)
+
+add_meta_data(
+    Aircraft.Battery.DISCHARGE_LIMIT,
+    meta_data=_MetaData,
+    historical_name={"GASP": None,
```
Comment (crecine) [inline aviary/variable_info/variable_meta_data.py:230]:
```suggestion
    historical_name={"GASP": 'INGASP.SOCMIN',
```

Comment (crecine) [reply aviary/variable_info/variable_meta_data.py:230]: proportion (0-1)

**Review by crecine — CHANGES_REQUESTED** [review]

File: aviary/models/engines/turbofan_28k_with_electric.deck
Comment (johnjasa) [reply aviary/models/engines/turbofan_28k_with_electric.deck:1]: I've made this change throughout to `electric_power_in`

File: aviary/subsystems/energy/battery_builder.py (hunk 1 above)
Comment (johnjasa) [reply aviary/subsystems/energy/battery_builder.py:13]: Maybe; are you asking if `get_mass_names` should instead be `mass_names`? If so, we should revisit more of the `get_` methods in subsystem builders in general; this would be a breaking change for users with external subsystems. Let me know if that's what you meant or if I'm misinterpreting.

File: aviary/subsystems/energy/battery_builder.py (hunk 2 above)
Comment (johnjasa) [reply aviary/subsystems/energy/battery_builder.py:19]: Done with "# Here, the efficiency variable is used as an overall efficiency for the battery"

File: aviary/subsystems/energy/battery_builder.py (hunk 3 above)
Comment (johnjasa) [reply aviary/subsystems/energy/battery_builder.py:56]: I removed the top commented out code; keeping the state_of_charge constraint there and commented out to show that it can be added. Added a comment to say this.

File: aviary/subsystems/energy/test/test_battery.py (hunk 4 above)
Comment (johnjasa) [reply aviary/subsystems/energy/test/test_battery.py:35]: Done!

File: aviary/subsystems/test/subsystem_tester.py (hunk 5 above)
Comment (johnjasa) [reply aviary/subsystems/test/subsystem_tester.py:271]: A user might add an external subsystem that integrates a rate source that doesn't exist in the subsystem. In this PR, we integrate the electric energy usage (which exists in core Aviary) to get the total battery energy used.

File: aviary/variable_info/variable_meta_data.py (hunk 6 above)
Comment (johnjasa) [reply aviary/variable_info/variable_meta_data.py:236]: Added 20%!

File: aviary/variable_info/variables.py (hunk 8 above)
Comment (johnjasa) [reply aviary/variable_info/variables.py:603]: Changed to `cumulative_electric_energy_used`

File: aviary/models/engines/turbofan_28k_with_electric.deck
Comment (johnjasa) [reply aviary/models/engines/turbofan_28k_with_electric.deck:1]: Constructed using the original `turbofan_28k` deck and a scalar on net thrust to get an arbitrary power draw; non-real source.

File: aviary/validation_cases/benchmark_tests/test_battery_in_a_mission.py (hunk 9 above)
Comment (johnjasa) [reply aviary/validation_cases/benchmark_tests/test_battery_in_a_mission.py:83]: Great point, I've added checks on the electricity and fuel used:

```
        electric_energy_used = prob.get_val(f'traj.cruise.timeseries.{av.Dynamic.Mission.CUMULATIVE_ELECTRIC_ENERGY_USED}', units='kW*h')
        fuel_burned = prob.get_val(av.Mission.Summary.FUEL_BURNED, units='lbm')

        # Check outputs
        assert_near_equal(electric_energy_used[-1], 42.5594728, 1.e-7)
        assert_near_equal(fuel_burned, 805.8963261, 1.e-7)
```

File: aviary/variable_info/variable_meta_data.py (hunk 14 above)
Comment (johnjasa) [reply aviary/variable_info/variable_meta_data.py:6223]: The latter, I've changed it to `'Total amount of electric energy consumed by the vehicle up until this point in the mission'`

File: aviary/subsystems/energy/battery_builder.py (hunk 1 above)
Comment (crecine) [reply aviary/subsystems/energy/battery_builder.py:13]:
```suggestion
    @property
    def get_mass_names(self):
```
so it can be accessed as `BatteryBuilder.get_mass_names` instead of `BatteryBuilder.get_mass_names()`, since no arguments are used

File: aviary/variable_info/variable_meta_data.py [hunk 17]
```diff
@@ -6213,18 +6214,18 @@
...
-    desc='Total amount of electric energy consumed by the vehicle across the mission',
+    desc='Total amount of electric energy consumed by the vehicle up until this point in the mission',
 )
 
 add_meta_data(
-    Dynamic.Mission.ELECTRIC_POWER,
+    Dynamic.Mission.ELECTRIC_POWER_IN,
     meta_data=_MetaData,
     historical_name={"GASP": None,
                      "FLOPS": None,
```
Comment (crecine) [inline aviary/variable_info/variable_meta_data.py:6231]: Maybe change the description from:
'Current electric power consumption of the vehicle, per single instance of each engine model'
to:
'Current electric power consumption of each engine'

File: aviary/variable_info/variables.py [hunk 18]
```diff
@@ -600,9 +600,9 @@ class Mission:
         DISTANCE_RATE = 'distance_rate'
         DRAG = 'drag'
         DYNAMIC_PRESSURE = 'dynamic_pressure'
-        ELECTRIC_ENERGY = 'electric_energy'
-        ELECTRIC_POWER = 'electric_power'
-        ELECTRIC_POWER_TOTAL = 'electric_power_total'
+        CUMULATIVE_ELECTRIC_ENERGY_USED = 'cumulative_electric_energy_used'
```
Comment (crecine) [inline aviary/variable_info/variables.py:603]: move this to the "C"s

File: aviary/subsystems/energy/battery_builder.py (hunk 10 above)
Comment (johnjasa) [reply aviary/subsystems/energy/battery_builder.py:29]: Good call, I've changed it so battery state of charge is constrained by default.

File: aviary/variable_info/variable_meta_data.py (hunk 17 above)
Comment (johnjasa) [reply aviary/variable_info/variable_meta_data.py:6231]: Done!

File: aviary/variable_info/variables.py (hunk 18 above)
Comment (johnjasa) [reply aviary/variable_info/variables.py:603]: Oh yeah, thanks!

File: aviary/subsystems/energy/battery_builder.py (hunk 1 above)
Comment (johnjasa) [reply aviary/subsystems/energy/battery_builder.py:15]: Got it, I like that, will make issue for future

File: aviary/subsystems/propulsion/engine_model.py [hunk 19]
```diff
@@ -40,7 +40,7 @@ class EngineModel(SubsystemBuilderBase):
     default_name = 'engine_model'
 
     def __init__(
-        self, name: str = None, options: AviaryValues = None, meta_data: dict = None,
+        self, name: str = None, options: AviaryValues = AviaryValues(), meta_data: dict = None,
     ):
         super().__init__(name, meta_data=meta_data)
         if options is not None:
```
[the remark this answers is not in the document]
Comment (johnjasa) [reply aviary/subsystems/propulsion/engine_model.py]: I see what you're saying, I changed the default back to `None` to handle this logic correctly

File: aviary/variable_info/variable_meta_data.py [hunk 20]
```diff
@@ -203,6 +203,123 @@
...
+)
+
+add_meta_data(
+    Aircraft.Battery.PACK_VOLUMETRIC_DENSITY,
+    meta_data=_MetaData,
+    historical_name={"GASP": None,
+                     "FLOPS": None,
+                     "LEAPS1": None
+                     },
+    units='kW*h/kg',
```
Comment (jkirk5) [inline aviary/variable_info/variable_meta_data.py:306]:
```suggestion
    units='kW*h/l',
```

File: aviary/subsystems/energy/battery_builder.py (hunk 1 above)
Comment (jkirk5) [reply aviary/subsystems/energy/battery_builder.py:15]: If it's a property I think changing to `mass_names` will be cleanest. There are probably other functions in the builder that could be properties too.

File: aviary/subsystems/energy/battery_builder.py [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+
+        return state_dict
+
+    def get_constraints(self):
+        constraint_dict = {
+            # Can add constraints here; state of charge is a common one in other battery applications
+            f'battery.{Dynamic.Mission.BATTERY_STATE_OF_CHARGE}':
+            {'type': 'boundary',
+             'loc': 'final',
+             'lower': 0.2},
```
Comment (jkirk5) [inline aviary/subsystems/energy/battery_builder.py:57]: I remember having a conversation about this earlier but we need to be able to constrain state of charge to the discharge limit, and we didn't have a good solution at the time. If there still isn't a clear path can you make an issue to tackle this later?

**Review by jkirk5 — APPROVED** [review]
Minor change request

File: aviary/subsystems/energy/battery_builder.py [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+                                                                        'defect_ref': 1e6,
+                                                                        'units': 'kJ',
+                                                                        'rate_source': Dynamic.Mission.ELECTRIC_POWER_IN_TOTAL,
+                                                                        'input_initial': 0.0}}
+
+        return state_dict
+
+    def get_constraints(self):
+        constraint_dict = {
+            # Can add constraints here; state of charge is a common one in other battery applications
```
Comment (crecine) [inline aviary/subsystems/energy/battery_builder.py:53]:
```suggestion
            # Can add constraints here; state of charge is a common one in many battery applications
```

File: aviary/subsystems/energy/test/test_battery.py [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+
+class TestBattery(av.TestSubsystemBuilderBase):
+
+    def setUp(self):
+        self.subsystem_builder = BatteryBuilder()
+        self.aviary_values = av.AviaryValues()
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (crecine) [inline aviary/subsystems/energy/test/test_battery.py:81]:
```suggestion
    unittest.main()

```

File: aviary/validation_cases/benchmark_tests/test_battery_in_a_mission.py [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+            f'traj.cruise.timeseries.{av.Dynamic.Mission.CUMULATIVE_ELECTRIC_ENERGY_USED}', units='kW*h')
+        fuel_burned = prob.get_val(av.Mission.Summary.FUEL_BURNED, units='lbm')
+
+        # Check outputs
+        assert_near_equal(electric_energy_used[-1], 42.5594728, 1.e-7)
+        assert_near_equal(fuel_burned, 805.8963261, 1.e-7)
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (crecine) [inline aviary/validation_cases/benchmark_tests/test_battery_in_a_mission.py:96]:
```suggestion
    unittest.main()

```

**Review by crecine — APPROVED** [review]
Minor suggestion, other than the change for the units

File: aviary/subsystems/energy/battery_builder.py (hunk 21 above)
Comment (johnjasa) [reply aviary/subsystems/energy/battery_builder.py:57]: Right on, I've made an issue for that!
